pkg/client/client.go

```diff
@@ -76,14 +80,71 @@ func NewHTTPClient(opts Options) *HTTPClient {
 		RetryWaitMin: retryMinWait,
 		RetryWaitMax: retryMaxWait,
 		RetryMax:     opts.MaxRetries,
-		CheckRetry:   retryablehttp.DefaultRetryPolicy,
+		CheckRetry:   RetryPolicy,
 		Backoff:      linearJitterRetryAfterBackoff,
 	}
 
 	client := retryClient.StandardClient()
 	return &HTTPClient{Client: client}
 }
 
+// RetryPolicy wraps retryablehttp.DefaultRetryPolicy and included additional logic:
+// - checks for specific errors that indicate a fall-back to the next download strategy
+// - checks for http.StatusBadGateway and http.StatusServiceUnavailable which also indicate a fall-back
+func RetryPolicy(ctx context.Context, resp *http.Response, err error) (bool, error) {
+	// do not retry on context.Canceled or context.DeadlineExceeded, this is a fast-fail even though
+	// the retryablehttp.ErrorPropagatedRetryPolicy will also return false for these errors. We can avoid
+	// extra processing logic in these cases every time
+	if ctx.Err() != nil {
+		return false, ctx.Err()
+	}
+
+	// While type assertions are not ideal, alternatives are limited to adding custom data in the request
+	// or in the context. The context clearly isolates this data.
+	consistentHashing, ok := ctx.Value(config.ConsistentHashingStrategyKey).(bool)
```

Comment:
out of interest, why make this consistent-hashing specific? if regular buffer mode hits a similar unrecoverable error, we could return ErrStrategyFallback and buffer mode just aborts. This is a change to current behaviour but if we're hitting 502 or 503 we're unlikely to retry our way out of that problem anyway.

(We can address this in another PR)